app_dart/lib/src/model/firestore/ci_staging.dart

```diff
@@ -0,0 +1,157 @@
+  static String documentNameFor({required RepositorySlug slug, required String sha, required String stage}) {
+    // Document names cannot cannot have '/' in the document id.
+    final docId = '${slug.owner}_${slug.name}_${sha}_$stage';
+    return '$kDocumentParent/ciStaging/$docId';
+  }
+
+  /// Lookup [Commit] from Firestore.
+  ///
+  /// Use [documentNameFor] to get the correct [documentName]
+  static Future<CiStaging> fromFirestore({
+    required FirestoreService firestoreService,
+    required String documentName,
+  }) async {
+    final Document document = await firestoreService.getDocument(documentName);
+    return CiStaging.fromDocument(ciStagingDocument: document);
+  }
+
+  /// Create [CiStaging] from a Commit Document.
+  static CiStaging fromDocument({
+    required Document ciStagingDocument,
+  }) {
+    return CiStaging()
+      ..fields = ciStagingDocument.fields!
+      ..name = ciStagingDocument.name!;
+  }
+
+  /// The remaining number of checks in this staging.
+  int? get remaining => int.tryParse(fields![kRemainingField]!.integerValue ?? '');
+
+  /// The total number of checks in this staging.
+  int? get total => int.tryParse(fields![kTotalField]!.integerValue ?? '');
+
+  /// Mark a [checkRun] for a given [stage] with [conclusion].
+  ///
+  /// Returns a [StagingConclusion] record or throws. If the check_run was
+  /// both valid and recorded successfully, the record's `remaining` value
+  /// signals how many more tests are running. Returns the record (valid: false)
+  /// otherwise.
+  static Future<StagingConclusion> markConclusion({
+    required FirestoreService firestoreService,
+    // required String documentNameString,
+    required RepositorySlug slug,
+    required String sha,
+    required String stage,
+    required String checkRun,
+    required String conclusion,
+  }) async {
+    final logCrumb = 'markConclusion(${slug.owner}_${slug.name}_${sha}_$stage, $checkRun, $conclusion)';
+
+    // Marking needs to happen while in a transaction to ensure `remaining` is
+    // updated correctly. For that to happen correctly; we need to as perform a
+    // read of the document in the transaction as well. So start the transaction
+    // first thing.
+    final docRes = await firestoreService.documentResource();
+    final transactionResponse = await docRes.beginTransaction(
+      BeginTransactionRequest(
+        options: TransactionOptions(
+          readWrite: ReadWrite(),
+        ),
+      ),
+      kDatabase,
+    );
+    final transaction = transactionResponse.transaction;
+    if (transaction == null) {
+      log.warning('$logCrumb: transaction was null');
+      throw '$logCrumb: transaction was null';
+    }
+
+    var newRemaining = -1;
+
+    late final Document doc;
+
+    // transaction block
+    try {
+      // First: read the fields we want to change.
+      final documentPath = documentNameFor(sha: sha, slug: slug, stage: stage);
```

Comment:
Are you also suggesting the document name be that as well?

flutter_flaux_engine_<sha>?

When I look at the document list, I kinda want:  slug/sha/stage, in case we add more stages later.

Updated the ordering here.

Reply:
It's just a habit of mine. I'm not sure if this can be applied consistently everywhere.